addons/mail/models/ir_ui_menu.py

```diff
@@ -0,0 +1,57 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import contextlib
+
+from odoo import api, models
+from odoo.exceptions import AccessError
+
+
+class IrUiMenu(models.Model):
+    _inherit = 'ir.ui.menu'
+
+    @api.model
+    def _get_best_backend_root_menu_id_for_model(self, res_model):
+        """Get the best menu root id for the given res_model and the access
+        rights of the user.
+
+        When a link to a model was sent to a user it was targeting a page without
+        menu, so it was hard for the user to act on it.
+        The goal of this method is to find the best suited menu to display on a
+        page of a given model.
+
+        Technically, the method tries to find a menu root which has a sub menu
+        visible to the user that has an action linked to the given model.
+        If there are more than one possibility, it chooses the menu root which
+        has the smallest parent_path (idea being that the module that introduce
+        the model will be installed first (smallest id) and the menu that leads
+        to the model should be in the first ones and not buried in the hierarchy).
+
+        :param str res_model: the model name for which we want to find the best
+            menu root id
+        :return (int): the best menu root id or None if not found
+        """
+        with contextlib.suppress(AccessError):  # if no access to the menu, return None
+            visible_menu_ids = self._visible_menu_ids()
+            # Try first to get a menu root from the model implementation (take the less specialized i.e. the first one)
+            menu_root_candidates = self.env[res_model]._get_backend_root_menu_ids()
+            menu_root_id = next((m_id for m_id in menu_root_candidates if m_id in visible_menu_ids), None)
+            if menu_root_id:
+                return menu_root_id
+
+            # No menu root could be found by interrogating the model so fall back to a simple heuristic
+            # Prefetch menu fields and all menu's actions of type act_window
+            menus = self.env['ir.ui.menu'].browse(visible_menu_ids)
+            self.env['ir.actions.act_window'].sudo().browse([
+                int(menu['action'].split(',')[1])
+                for menu in menus.read(['action', 'parent_path'])
+                if menu['action'] and menu['action'].startswith('ir.actions.act_window,')
+            ]).filtered('res_model')
+
+            # Return the root menu of the visible menu with the smallest id and act_window targeting the model
+            menu_sudo = min((
```

Comment:
Now that we have paths on actions since ``odoo/odoo#142007`` shouldn't we prioritize them if it is set, as it indicates a "major" action ?

Reply:
Yes indeed, it is a good idea. I have added a preference function that depends on the menu and the action which prioritize the menu linked to an action having a path. And I have modified the test accordingly.